### ISSUE TITLE
Develop a linter that looks for non-prod code in dart2js output

### ISSUE THREAD

**Issue by yjbanov** [angular/angular#3440]
We currently have little idea what ends up in dart2js output and why. Moreover, it would be nice to protect ourselves from future regressions (accidental inclusion of non-prod code in production output).

## Proposal

Introduce the following annotations that can be placed on libraries, classes, functions and methods:

### `@Prod`

Annotates code that's meant to run in production. The sole field in the annotation describes the reason. Example:

```typescript
@Prod('uses browser DOM APIs')
export class DomRenderer {
  ...
}

@Prod('manipulates DOM nodes')
function moveNodesAfterSibling(sibling, nodes) {
  ...
}
```

### `@NonProd`

Annotates code that should not run in production. The sole field in the annotation describes the reason. Example:

```typescript
@NonProd('runs in transformer')
export class ViewSplitter implements CompileStep {
  ...
}
```

### Checker

The build checker itself would be a transformer that can be installed in the `pubspec.yaml` file of an application _after_ the `$dart2js` transformer. It will load the generated `--dump-info` JSON file and verify that none of the code is marked as `@NonProd`.

### Annotating ng2 code

For this to work we need to make sure the code is well annotated. To that end we'll introduce a Travis build checker that verifies that every `.ts` and `.dart` file in the `angular/angular` repo meets one of the following conditions:

- The file is annotated
- Every top-level member in the file is annotated

To avoid big bang update of the entire repo we'll annotate code in two phases:

#### Phase 1: non-failing warnings & clean up

During this phase Travis will produce warnings but it will not fail the build.

#### Phase 2: non-annotated code fails the build

Travis fails the build if it finds code that's not annotated. This will be the final state of things.

**Comment by mhevery** [comment]
1) I don't think this is P1, drooping to P3. No one is complaining of this, its nice to have.

2) I don't think annotations are the way to go, as they would bloat the TS code. We have to come up with a way of annotating things without adding to the payload.

**Comment by yjbanov** [comment]
#1 The complaint is in the form of "hello world" is 400kb
#2 We can strip these annotations during when publishing to NPM and elsewhere

**Comment by sigmundch** [comment]
I am not sure what's the most effective way to make progress on this, but I agree with @yjbanov that we need to start looking at what's causing angular2's code size to grow. In particular, I agree that #3315 is P1. I think opt-in or opt-out annotations can help guide the process, but there might be other ways. If you like, we can schedule a time to brainstorm some ideas.

I am separately looking at general ways to make it easier to track down source code culprits in dart2js's dump-info, partially by using code-coverage information and by using callgraph dependencies to figure out what's pulling a lot of code. To make it easier to analyze this kind of data, it would be really useful to have framework specific hints that can narrow down the search. This is where I imagine that labeling what portions are never expected to be part of the production app (as suggested with `@prod` above) can come in handy.

**Comment by yjbanov** [comment]
Making it P1. We need this (hopefully) before we start removing dynamic pieces from runtime. Otherwise it is too easy to pull them back in without knowing it.

**Comment by sigmundch** [comment]
Another though - could we have a notion of `@DevInjectable`? That way we can automatically exclude from the transformed code anything that would have been injected for development.

**Comment by yjbanov** [comment]
@sigmundch would combining `@Injectable` with `@NonProd()` not work?

There is a bigger issue with transformers that we haven't addressed yet. Namely that they pull everything into ng_deps files, used or not used. Perhaps the transformer should have a prod mode in which it crawls the sources looking for things that are actually used in the app. We could, of course, also ask people to break up their code into smaller libraries (e.g. not over-use `part of`). We're still kind of defeating the tree shaking capability.

**Comment by kegluneq** [comment]
Right, we've discussed some potential solutions for the tree shaking issue in the [design doc](https://docs.google.com/document/d/1Oe7m96QnOrilxpH1B5o9G_PnfBGovhH-n_o7RU6LYII/edit#)

**Comment by mhevery** [comment]
@yjbanov Sorry for dropping this to P3 the second time around. I think I was doing some bulk operations on issues and this issue got coughed in the bulk. 

I agree that figuring out how to shrink size of Hello World is P1, it just is not clear to me what specific actions need to be taken. I think we should focus on understanding the problem before we start guessing on the solutions.

**Comment by naomiblack** [comment]
obsolete. we are addressing hello world size via codegen + treeshaking.
